Convert all DID Method entries to JSON files.

msporny opened the pull request:
This PR is the first in a series to convert all DID Method entries to JSON files that contain links to information about the DID Method. This PR is purely additive, it only adds files that are not used at the moment, a future PR will use these files to auto-generate the table(s) of DID Methods.

This PR also removes the "PROVISIONAL" status marker, keeping the "DEPRECATED" and "WITHDRAWN" markers.


<!--
    This comment and the below content is programatically generated.
    You may add a comma-separated list of anchors you'd like a
    direct link to below (e.g. #idl-serializers, #idl-sequence):

    Don't remove this comment or modify anything below this line.
    If you don't want a preview generated for this pull request,
    just replace the whole of this comment's content by "no preview"
    and remove what's below.
-->
***
<a href="https://pr-preview.s3.amazonaws.com/w3c/did-spec-registries/pull/353.html" title="Last updated on Nov 1, 2021, 4:22 PM UTC (be8631c)">Preview</a> | <a href="https://pr-preview.s3.amazonaws.com/w3c/did-spec-registries/353/29684c9...be8631c.html" title="Last updated on Nov 1, 2021, 4:22 PM UTC (be8631c)">Diff</a>

TallTed commented on methods/key.json:
```diff
@@ -0,0 +1,12 @@
+  "name": "key",
+  "description": "",
+  "status": "",
+  "registry": "Ledger independent DID method based on public/private key pairs",
```
For consistency
```suggestion
  "registry": "Ledger-independent DID method based on public/private key pairs",
```

TallTed reviewed (approved):
One tweak for internal consistency. Otherwise, appears to do what it says on the tin.

OR13 commented on methods/3.json:
```diff
@@ -0,0 +1,12 @@
+  "description": "",
+  "status": "",
+  "registry": "Ceramic Network",
+  "contact": "<a href=\"mailto:oed@3box.io\">Joel Thorstensson</a>",
```
<script>alert(1);</script>

OR13 replied:
prefer json to not include XSS vectors.

OR13 replied:
> At present, we allow HTML markup in all entries (this is nothing new, we've operated like this for a while).

This PR is the first time we have added JSON files which contain html markup to the registry, and we should not start by encouraging dom based xss via respec.

is contact always an email? why not call it "email" and make it an email?

msporny replied:
> is contact always an email? why not call it "email" and make it an email?

No, it isn't. Sometimes it's an email. Sometimes it's a website. Sometimes it's just a name. Sometimes it's just the name of a company. So, what of those options are we going to force people to do (because they all seem like valid options, IMHO).

OR13 replied:
ok thats fail, I think its easier to express that as optional fields.... Like this:

```yaml
title: DID Method Registry Entry
description: This schema defines the shape of did method registrations in the did spec registries
type: object
required:
  - status
  - name
  - specification
properties:
  status:
    type: string
    enum: ["accepted", "withdrawn", "deprecated"]
  name:
    type: string
    example: "did:example"
  description:
    type: string
    example: "did:example is reserved for examples."
  specification:
    type: string
    example: "https://workday.github.io/work-did-method-spec/"
  contactName:
    type: string
    example: "Example Corp. Support"
  contactEmail:
    type: string
    example: "support@example.com"
  contactWebsite:
    type: string
    example: "example.com"
  verifiableDataRegistry:
    type: string
    example: "Bitcoin / Ethereum + IPFS / Hyperledger"
  rubricEvaluation:
    type: string
    example: ""
additionalProperties: false
```

OR13 replied:
`additionalProperties: false` <- this is key.

msporny replied:
I'm reworking the schema you have above into the conversion script (I don't think we need maxLength, but *shrug* -- if you want it go ahead and put it in there). Bonus points if you want to work the JSON Schema above into a Github merge test for all PRs in the methods/ directory.

msporny replied:
I implemented the schema above in the latest PR commits. The only change I didn't take was the "accepted" language... I chose to use "registered" instead.

We may not even want to have "registered", since it's obvious that it's registered if it's in the DID Spec Registries. So, it's either blank or "withdrawn"/"deprecated". We may even want to collapse both of those entries into "deprecated".

OR13 reviewed (changes requested):
@msporny thanks for this PR.

I have a few concerns, listing them in priority order here:

### 1. Preventing Double Work.

If we accept this PR, it should be complete with instructions for new registrations, and functional, in that there is only a single entry related to existing registry entries in the repo after merging.

Otherwise, we will have folks adding to the "old registry" making more work for maintainers of the "new registry"...

### 2. Documenting All JSON Members

There should be a description of each JSON member, and its requirements.

Allowed characters, min max length, required or optional, etc...

### 3. Document New Registration Process

Instructions should be provided so we can adopt the new process completely, and close all open PRs related to the old process.

### 4. Can we do this via code first?

is there a script that is used to generate these files?

I prefer to review a script first, not a bunch of generated files... easier to adjust the output as desired.

This would make it much easier to review the intention behind the PR.


## Next Steps

I can help document the proposed JSON members and help on a script to generate these files.

Can we remove the new files being added and instead add the script, json requirements and new registration requirements, and then generate the data at the end, once we have agreed to the new structure?

OR13 commented:
Here are some constraints on the registry members and their descriptions: https://github.com/w3c/did-spec-registries/pull/353#discussion_r740440602

msporny commented:
> 4. Can we do this via code first?

Code is here:

https://gist.github.com/msporny/1fdcb0b98050b849614e1c9fd6a341de

> 1. Preventing Double Work.
> If we accept this PR, it should be complete with instructions for new registrations, and functional, in that there is only a single entry related to existing registry entries in the repo after merging.
> Otherwise, we will have folks adding to the "old registry" making more work for maintainers of the "new registry"...

No, for two reasons:

1. The PR is meant to be coupled with other PRs for instructions for new registrations (feel free to work on that PR, please).
2. There is a conversion script, so anything merged in the interim is easy to convert out of the main index.html until we replace it with the ReSpec autogenerator (which I plan to get done soon-ish).

There is a plan to do what you want, but -1 to bundling them all into a single PR. All this PR does is the conversion. A separate PR will do the instructions for new registrations. Yet another separate PR will do the ReSpec auto-generation of the DID Method table. That's the plan so that we can have PRs that do one thing so it's easier to collaborate and merge on each subunit of work.

> 2. Documenting All JSON Members
> There should be a description of each JSON member, and its requirements.

Please document this and put it into a PR in a way that is acceptable to you so that we can review.

> 3. Document New Registration Process
> Instructions should be provided so we can adopt the new process completely, and close all open PRs related to the old process.

Same as above... separate PR, please.

If you choose to work on the two separate PRs above, please let me know so we don't duplicate work.

OR13 commented:
@msporny based on your feedback, let me take a stab at a separate PR, consider the ball in my court.

OR13 commented:
Lets close this PR, and continue to discussion on: https://github.com/w3c/did-spec-registries/pull/356

This PR adds a too many generated files, and does not meet the acceptance criteria for a transition of registry format I listed above.

OR13 commented:
@msporny  can we close this? your PR https://github.com/w3c/did-spec-registries/pull/357 seems a better starting place.

msporny commented:
> @msporny can we close this? your PR #357 seems a better starting place.

PR #357 is based on top of this one. :)

mprorock commented:
I would like to also see a an optional link to an example in addition to the optional link to to the rubric eval in the schema for the json entries

msporny commented:
> I would like to also see a an optional link to an example in addition to the optional link to to the rubric eval in the schema for the json entries

Can we do this in another PR that just focuses on "fields we want to add"?

msporny commented:
Based on feedback from the last telecon, I have:

* Updated this PR to only cover converting the current HTML-based entry method that we have to a JSON-based entry method. 
* Removed any new proposed fields, which will be handled in a future PR (that proposes the new field(s) and modifies all JSON files to include the new field(s)).
* Temporarily included the tooling to do the conversion, which will be removed before merging to main.
* Renamed "provisional" to "registered".

I believe that removes every controversy in this PR that was identified during the last telecon. I believe this PR is now ready for review and merge.

OR13 reviewed (changes requested):
This PR must be merged along with:

- https://github.com/w3c/did-spec-registries/pull/361 
- https://github.com/w3c/did-spec-registries/pull/360
- https://github.com/w3c/did-spec-registries/pull/357

Or not at all.


I think this PR should be merged first, of the series, so as long as any of the PRs above has change requests, this one should not be merged.

msporny commented:
> I think this PR should be merged first, of the series, so as long as any of the PRs above has change requests, this one should not be merged.

Agree.

iherman commented:
The issue was discussed in a [meeting](https://www.w3.org/2019/did-wg/Meetings/Minutes/2021-11-04-did-topic#section2-7) on 2021-11-04

- no resolutions were taken

<details><summary><i>View the transcript</i></summary>

#### 2.7. Convert all DID Method entries to JSON files. (pr did-spec-registries#353)

_See github pull request [did-spec-registries#353](https://github.com/w3c/did-spec-registries/pull/353)._




**Orie Steele:** This is manu's draft, PR 353..  

**Manu Sporny:** PR 353 attempts to turn all the registry entries into JSON files.  
… it changes everything into Orie's YAML file format.  
… it renames "Provisional" to "Registered" and retains "Deprecated" and "Withdrawn".  

> *Manu Sporny:* [https://github.com/w3c/did-spec-registries/pull/353/files#diff-c68755f9bc842d714659a770fdade9d566c926890c48f9448152b57d70f7b14d](https://github.com/w3c/did-spec-registries/pull/353/files#diff-c68755f9bc842d714659a770fdade9d566c926890c48f9448152b57d70f7b14d).

**Manu Sporny:** that's an example of what a registration looks like..  
… it allows for a description field.  
… it breaks out name and email.  
… and it has links to implementation, test suite, and other self-asserted info.  
… this PR does not address any review of the DID method specification.  
… the PR itself is just to change the overall composition of the table.  
… the spec, the name, and the status are required; all others are optional.  
… the purpose of this PR is just to redo the format of the current table.  
… other PRs can then tackle the rendering challenge.  

**Orie Steele:** Most of what Manu said, I agree with. This PR adds some new empty fields to every entry, it also duplicates registry entries if we were to merge it w/o additional requirements. My two primary objections -- duplicates data w/o registration process, adds fields that don't exist today..  
… I did attempt to address in future PRs. We might want to talk about what content should look like... registration changes in one shot..  

> *Kristina Yasuda:* +1 Orie.

**Orie Steele:** if I got it to the point, we could merge it in one shot..  

**Drummond Reed:** Is fastest thing to do to talk about convergence plan?.  

> *Orie Steele:* ^ yeah, lets look at my PR and then talk about differences.

**Drummond Reed:** Almost seems the same..  

**Manu Sporny:** let's just talk about the differences.  
… my opinions become stronger if I have to build/maintain the system.  
… my suggestion is to talk about the end-result and work backwards.  
… let's make sure we clarify what the final result is, then the editors can figure out how to accomplish it.  
… for example, is there a status field or not.  
… do we want to provide a description field?.  
… do we or do we not make value judgments?.  
… the mechanics matter only to the editors, but the fields we see matter to everyone.  

> *Orie Steele:* +1 manu.

**Manu Sporny:** my suggestion is that we go "top down", deciding on what we want to show up, then let the editors tackle the tech details.  

**Orie Steele:** I agree, my goal with these PRs was to have what folks see at the end of what people see w/o the word provisional in it..  
… MikeJ's PR accomplishes that goal as well -- rendering the same information we have today w/o misleading value judgments in the table... don't add any new fields, preserve existing talbe structure, rename provisional to registered -- first thing is what folks should see is not the word provisional, existing registry entries w/ no additional changes..  
… However, folks will need to see documentation on how to register a DID Method, it's not how process was before -- they will need instructions on how to add JSOn files -- my PR covers those instructions. Those are also a requirement, to summarize -- what folks should see is same information they see today w/o provisional showing up... they can read, update DID Method..  

> *Orie Steele:* I will not approve a PR that mixes adding new fields with removing provisional.

**Drummond Reed:** I understand, we're merging two different perspectives -- I hear Orie saying "let's do this one thing for what people see in the table" -- add new registration process -- Manu saying "Do we want to add additional fields?" and there is a 4th decision -- are we going to make any value judgement at all? My proposal is only one..  

> *Orie Steele:* +1 manu... either we keep the existing process and remove provisional.... or we update the process, document it, and remove provisional.

**Drummond Reed:** What do we want to see? Are we making any value judgement about specification? I'm still proposing baseline value judgement, if you want to call it that. Conformance judgement..  

> *Kristina Yasuda:* I think we agree not making any value judgement in the registry.  
> *Kristina Yasuda:* making a decision to include a DID method to a registry is *different* from making value judgments.

**Manu Sporny:** the issue I have with Orie's PR is that it puts multiple issues together in one PR.  
… if we just want to remove "Provisional", then let's just modify Mike's PR.  
… but the other stuff still remains.  

> *Brent Zundel:* +1 to pulling in Mike's PR as a first step.

> *Kristina Yasuda:* +1 to PRs per issue.

**Manu Sporny:** I'd rather talk about those in "clean PRs".  

> *Orie Steele:* let see if we can get Mikes PR merged on the call then..

**Manu Sporny:** one that talks about the fields.  

> *Charles Lehner:* Removing withdrawn or deprecated entries might be considered as a separate change also.

**Manu Sporny:** a second one that renders that data.  

> *Orie Steele:* -1 to splitting generating and rendering... but also -1 to doing any of that until Mike's PR is merged.

**Manu Sporny:** the third PR is specifically what are the new registration requirements.  
… if we don't do it that way, it's going to be all tangled up in one PR that will take forever.  
… so if we want to get rid of "Provisional", we should merge Mike's PR.  

> *Brent Zundel:* but those issues don't retain the content.

> *Kristina Yasuda:* If withdrawn and deprecated is not defined in the spec, Mike's PR should be merged as-is.

**Orie Steele:** We have statuses -- I do agree with Manu, changing the process and ability to retain registry entries is hard - in favor of merging Mike's PR given that we have open issues to track what withdrawn/deprecated meant, so Mike's PR is acceptable to merge. We haven't lost any info, we can do so when we update process to how to handle deprecation, how to handle withdrawn. If it's possible -- we don't want to commingle that issue..  

> *Orie Steele:* all status are not defined...  
> *Orie Steele:* all undefined status should be removed first...  
> *Orie Steele:* then defined.
> *Orie Steele:* then added back if wee can get agreement...

> *Kristina Yasuda:* +1 Orie.

**Drummond Reed:** Sounds like we're converging -- to address what Kristina said -- withdrawn/deprecated not defined in DID Spec, provisional wasn't defined, status column exists, no official process -- inherited from CCG..  
… if we agree on that, would like to second manu's proposal, 3 separate PRs... what is the data, how are we going to render it, what is the registration process?.  
… That's where we can tackle conformance judgement about spec or not..  

**Orie Steele:** Comment on where we're converging, let's focus on Mike's PR. Manu said he'd object to merging if we can't preserve existing statuses..  

**Manu Sporny:** I'm objecting that we should not just remove any registry information..  
… the latest PRs still preserve that info.  
… there is a reason that info is in the registry.  
… so the issue I'm concerned about is deleting info without checking with the registrant.  
… if we want to reach out to them, we might be able to make that decision.  
… that's the reason I wanting to keep it somewhere, such as a comment field.  

> *Kristina Yasuda:* I don't agree with that. We have a valid, logical reason to remove and once we define the provisional, etc. we can re-add withdrawal etc..

**Manu Sporny:** so "change the status column to a Comments column" and then add those comments.  

> *Justin Richer:* -1.

**Drummond Reed:** I know Dave Huseby would send a torpedo -- he was emphatic that status of `did:git` was withdrawn and he doesn't want anyone to misinterpret that..  
… We'd have a big issue there. I'd be surprised if that wasn't the case. Let's not lose that information, we're going to end up replacing it with something, we just need to figure out how to go through this transition. Interested in Justin's -1..  

**Kristina Yasuda:** This will not set a bad precedent, valid logical reason to remove, not removing for the sake of removing. We could put a short note, do not use [did:git](did:git) -- make sure Dave doesn't have an issue -- once we handle the definition of withdrawn and deprecated, we can always clarify that. Incremental approach, most logical approach, merge Mike's PR as is..  

> *Drummond Reed:* IMHO the info about the Deprecated and Withdrawn methods can be retained in notes outside the table..

> *Kristina Yasuda:* we can just add a text saying do NOT use `did:github` instead of labeling it "provisional".

> *Brent Zundel:* +1 to Drummond.

**Justin Richer:** The more you push a registry to just be free text, the more it becomes not like a registry -- comment, read the bits, aspirational at best, disaster at worst... purpose of registry is get to code points that mean very specific things. That's the main point of a registry..  

> *Orie Steele:* +1 justin, we ought not to fill the registry with fluff or undefined terms..

**Justin Richer:** It is purview of the registry to manage contents, it is up to registry maintainers to make the determination, make best effort to reach out to people -- tell people you're changing it... don't dance around -- I can't change that text, it's not mine -- absurd, not taking control of data that this registry is meant to control..  

> *Kristina Yasuda:* +1 Justin.

**Drummond Reed:** I think it's fine, I do agree with Justin, registry controls what's there -- maintain fidelity with what's registered, status has never been fully defined, fine to move information that two methods have status information outside of table -- get that done, can start on steps manu has started earlier..  

> *Justin Richer:* to be clear, I like a status column but with clearly defined values tied back to a spec definition.

_See github pull request [did-spec-registries#351](https://github.com/w3c/did-spec-registries/pull/351)._




**Manu Sporny:** Merge the PR, go ahead and get it done even though I object.  

**Orie Steele:** Ok, merged..  

> *Orie Steele:* process.

**Drummond Reed:** Let's agree on next agenda -- process, conformance, not necessarily sequential... what can we get done between now and next call..  

</details>

iherman commented:
The issue was discussed in a [meeting](https://www.w3.org/2019/did-wg/Meetings/Minutes/2021-11-09-did#section4-1) on 2021-11-09

- no resolutions were taken

<details><summary><i>View the transcript</i></summary>

#### 4.1. Convert all DID Method entries to JSON files. (pr did-spec-registries#353)

_See github pull request [did-spec-registries#353](https://github.com/w3c/did-spec-registries/pull/353)._




**Manu Sporny:** So, the first one that we're talking about is PR #353..  
… All this PR does is that it takes the existing table of DID methods and converts it into JSON. It doesn't add new fields, it doesn't add pointers to rubric or anything. It's just the table we have right now converted to JSON. Shouldn't be controversial..  
… In the future, we can have other fields with links to rubric, examples, things like that -- but that's not in this PR..  
… Any questions?.  

> *Drummond Reed:* No, that's clear..

</details>

msporny commented:
Editorial, multiple reviews, changes requested and made, no remaining objections, merging.
